native/swift/Sources/wordpress-api-cache/DatabaseChangeNotifier.swift

```diff
@@ -0,0 +1,294 @@
+/// }
+///
+/// // Start observing a collection
+/// DatabaseChangeNotifier.shared.startObserving(observableCollection) { updateHook in
+///     print("Collection updated: \(updateHook)")
+/// }
+///
+/// // Stop observing when done
+/// DatabaseChangeNotifier.shared.stopObserving(observableEntity)
+/// DatabaseChangeNotifier.shared.stopObserving(observableCollection)
+/// ```
+///
+/// - Note: Observers are automatically replaced if you start observing the same entity/collection multiple times.
+///         The previous observer will be cleaned up automatically.
+public final class DatabaseChangeNotifier: DatabaseDelegate, Sendable {
+
+    /// The shared singleton instance of the database change notifier.
+    public static let shared = DatabaseChangeNotifier()
+
+    /// Internal storage for notification observer tokens.
+    private let tokenStore = TokenStore()
+
+    private init() {}
+
+    /// Called by the database layer when an update occurs.
+    ///
+    /// This method posts notifications to both entity-specific and collection-wide observers
+    /// based on the update information provided in the hook.
+    ///
+    /// - Parameter updateHook: Information about the database update, including table name and row ID.
+    public func didUpdate(updateHook: WordPressAPIInternal.UpdateHook) {
+        NotificationCenter.default.post(name: entityNotificationName(for: updateHook), object: nil, userInfo: [
+            "updateHook": updateHook
+        ])
+
+        NotificationCenter.default.post(name: collectionNotificationName(for: updateHook), object: nil, userInfo: [
+            "updateHook": updateHook
+        ])
+    }
+
+    /// Starts observing changes to a specific database entity.
+    ///
+    /// Registers a callback to be invoked whenever the specified entity is updated in the database.
+    /// If an observer is already registered for this entity, it will be replaced and the previous
+    /// observer will be automatically cleaned up.
+    ///
+    /// - Parameters:
+    ///   - object: The entity to observe.
+    ///   - callback: A closure called when the entity is updated. The closure receives an `UpdateHook`
+    ///               containing information about the database change.
+    ///
+    /// - Note: You must call `stopObserving(_:)` when you're done observing to prevent memory leaks.
+    public func startObserving(
+        _ object: ObservableEntity,
+        callback: @escaping @Sendable (UpdateHook) -> Void
+    ) {
+        let token = NotificationCenter.default.addObserver(
+            forName: notificationName(for: object),
+            object: nil,
+            queue: nil
+        ) { notification in
+            guard let updateHook = notification.userInfo?["updateHook"] as? UpdateHook else { return }
+            callback(updateHook)
+        }
+
+        tokenStore.add(token, for: object.entityId) {
+            NotificationCenter.default.removeObserver($0)
+        }
```

Comment:
I can see this hand-writen observer implementation is necessary for Linux, where Combine is not available. But for Apple platforms, maybe we should use Combine.Publisher, which can turn into AsyncSequence (thus will work well with the async code)?